prog/rand.go

```diff
@@ -361,64 +367,54 @@ func (r *randGen) createResource(s *state, res *ResourceType, dir Dir) (arg Arg,
 		}
 		sort.Strings(all)
-		kind = all[r.Intn(len(all))]
-	}
-	// Find calls that produce the necessary resources.
-	metas0 := r.target.resourceCtors[kind]
-	// TODO: reduce priority of less specialized ctors.
-	var metas []*Syscall
-	for _, meta := range metas0 {
-		if s.ct.Enabled(meta.ID) {
-			metas = append(metas, meta)
+		kind1 := all[r.Intn(len(all))]
+		metas1 := r.enabledCtors(s, kind1)
+		if len(metas1) != 0 {
+			// Don't use the resource for which we don't have any ctors.
+			// It's fine per-se because below we just return nil in such case.
+			// But in TestCreateResource tests we want to ensure that we don't fail
+			// to create non-optional resources, and if we spoof a non-optional
+			// resource with ctors with a optional resource w/o ctors, then that check will fail.
+			kind, metas = kind1, metas1
 		}
 	}
 	if len(metas) == 0 {
-		return res.DefaultArg(dir), nil
+		// We may not have any constructors for optional input resources because we don't disable
+		// syscalls based on optional inputs resources w/o ctors in TransitivelyEnabledCalls.
+		return nil, nil
 	}
-
 	// Now we have a set of candidate calls that can create the necessary resource.
-	for i := 0; i < 1e3; i++ {
-		// Generate one of them.
-		meta := metas[r.Intn(len(metas))]
-		calls := r.generateParticularCall(s, meta)
-		s1 := newState(r.target, s.ct, nil)
-		s1.analyze(calls[len(calls)-1])
-		// Now see if we have what we want.
-		var allres []*ResultArg
-		for kind1, res1 := range s1.resources {
-			if r.target.isCompatibleResource(kind, kind1) {
-				allres = append(allres, res1...)
-			}
-		}
-		sort.SliceStable(allres, func(i, j int) bool {
-			return allres[i].Type().Name() < allres[j].Type().Name()
-		})
-		if len(allres) != 0 {
-			// Bingo!
-			arg := MakeResultArg(res, dir, allres[r.Intn(len(allres))], 0)
-			return arg, calls
-		}
-		// Discard unsuccessful calls.
-		// Note: s.ma/va have already noted allocations of the new objects
-		// in discarded syscalls, ideally we should recreate state
-		// by analyzing the program again.
-		for _, c := range calls {
-			ForeachArg(c, func(arg Arg, _ *ArgCtx) {
-				if a, ok := arg.(*ResultArg); ok && a.Res != nil {
-					delete(a.Res.uses, a)
-				}
-			})
+	// Generate one of them.
+	meta := metas[r.Intn(len(metas))]
+	calls := r.generateParticularCall(s, meta)
+	s1 := newState(r.target, s.ct, nil)
+	s1.analyze(calls[len(calls)-1])
+	// Now see if we have what we want.
+	var allres []*ResultArg
+	for kind1, res1 := range s1.resources {
+		if r.target.isCompatibleResource(kind, kind1) {
+			allres = append(allres, res1...)
 		}
 	}
-	// Generally we can loop several times, e.g. when we choose a call that returns
-	// the resource in an array, but then generateArg generated that array of zero length.
-	// But we must succeed eventually.
-	var ctors []string
-	for _, meta := range metas {
-		ctors = append(ctors, meta.Name)
+	sort.SliceStable(allres, func(i, j int) bool {
```

Comment:
Why sort it if we anyway take a random element from it?

Reply:
This is required for determinism and deterministic code coverage.
First, we want to be able to run a test with the same seed and get the same results each time.
Second, we generate coverage reports and if coverage is non-deterministic we get random fluctuations on PRs not related to changes themselves (e.g. docs-only PR changes prog coverage).